[5.4] Allow schedule:run to run as a daemon

tback opened the pull request:
Adds an option --loop so the scheduler will run forever. Triggering events once every 60 seconds by default.
Adds an option --interval so the interval can be adapted.

I added this option because it is handy in kubernetes. By running the event scheduler in a loop it can be a well behaved pod.

This is my first contribution to this project. I read the contribution guidelines, but I might have missed something.

tback committed:
Allow schedule:run to run as a daemon.

Adds an option --loop so the scheduler will run forever. Triggering
events once every 60 seconds by default.
Adds an option --interval so the interval can be adapted.

GrahamCampbell commented on src/Illuminate/Console/Scheduling/ScheduleRunCommand.php:
```diff
@@ -41,12 +42,49 @@ public function __construct(Schedule $schedule)
+                break;
+            }
+
+            //pause for a minimum of one second.
```
Please add a space character between the slashes and the comment body.

tback replied:
Fixed in adc8a0dd99a7dcb8d3191a390e3be234eab039c4

fernandobandeira commented on src/Illuminate/Console/Scheduling/ScheduleRunCommand.php:
```diff
@@ -41,12 +42,49 @@ public function __construct(Schedule $schedule)
     {
+        $loop = $this->option('loop');
+        $interval = $this->option('interval');
+        if (! is_int($interval) && $interval <= 0) {
```
Is this necessary? If the $interval is <= 0  then it will default to 1 anyway.

And also shouldn't it use || instead of &&?

tback replied:
You're right. That was wrong. `is_int` on a string confused me. 

Fixed it.

tback committed:
Remove unnecessary and broken bounds check

tback committed:
Clarify options of schedule:run

Rename option loop to daemon. Remove redundant information about default
interval.

tback committed:
Use camelCase for local variable

tback committed:
Return doc

tback committed:
Variable visibility

tback committed:
Throw exception in case of invalid argument

tback committed:
Code style

tback commented on src/Illuminate/Console/Scheduling/ScheduleRunCommand.php:
```diff
@@ -41,12 +42,49 @@ public function __construct(Schedule $schedule)
+        $interval = $this->option('interval');
+        if (!is_int($interval) && $interval <= 0){
+            $this->error('Interval must be an 1 <= integer <= 86400');
+            exit(1);
```
Fixed in 2f9e6c09d110a7bed32c0ae3fa6fdcc07043e369

tback commented on src/Illuminate/Console/Scheduling/ScheduleRunCommand.php:
```diff
@@ -41,12 +42,49 @@ public function __construct(Schedule $schedule)
+            }
+
+            //pause for a minimum of one second.
+            $sleep_time = max(1, $interval - (time() - $start));
```
Fixed in a6a86a84018cac78a203b97b9daa8f788d0c818f

tback commented on src/Illuminate/Console/Scheduling/ScheduleRunCommand.php:
```diff
@@ -41,12 +42,49 @@ public function __construct(Schedule $schedule)
+    /**
+     * Trigger due events.
+     */
+    private function doScheduleDueEvents()
```
Fixed in 9b1c31faf593252cac39609e11a54e8d44500f68

tback commented on src/Illuminate/Console/Scheduling/ScheduleRunCommand.php:
```diff
@@ -67,4 +105,5 @@ public function fire()
             $this->info('No scheduled commands are ready to run.');
         }
     }
+
```
Fixed in 40ed862fc9b4243e33c8dbaee40c3e82e542c128. This was force push integrating changes requested by style ci.

tback commented:
Requested changes should be done. Sorry for the mess.

taylorotwell commented:
This may lead to some undefined behavior. What if one of the tasks takes 2 minutes to run? Now you're actually running the scheduler every two minutes instead of every 1 minute, etc.

tback committed:
Stricter daemon mode

Remove interval option. Log error in case schedule:run exceeds 60
seconds.

tback commented:
@taylorotwell 
I wasn't aware schedule:run has those strict requirements. I therefore removed the interval option, explained the requirement in --daemon help and log an error if this ever happens.

I use schedule:run only to dispatch jobs to workers, so this should not be a problem. People who use schedule:run for long running job will probably have a hard time setting up a reliable daemon with systemd or supervisor as well. I reckon those people will stay with cron anyway.

taylorotwell commented:
Probably a better way to do this entirely and still recreate the exact behavior of how cron would work would be to fire off `php artisan schedule:run &` so that the process is started in the background as a new process, then sleep for 60 seconds, then do it again, etc.

tback commented:
That would create operations problems in it's own. I'm quite happy with my solution for the time being. I just released it as a package: https://github.com/tback/laravel-scheduledaemon
